### PR TITLE
(DRAFT) Drivetrain Testing Improvements (NOT READY YET)I

### PR DESCRIPTION
**Pull request by Isopod00**
Ready to merge

### CODE REVIEW

File: config/drivetrain_config.yaml [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
-        HALF_WHEEL_BASE: 0.7127  # HALF of the wheelbase in meters # TODO: Verify this on the final robot
-        HALF_TRACK_WIDTH: 0.36585  # HALF of the trackwidth in meters # TODO: Verify this on the final robot
-        STEERING_MOTOR_GEAR_RATIO: 20 # TODO: Ask mechanical what this value will be on the real robot
-        FRONT_LEFT_MAGNET_OFFSET: 0 # TODO: Calibrate this on the final robot
-        FRONT_RIGHT_MAGNET_OFFSET: 0 # TODO: Calibrate this on the final robot
-        BACK_LEFT_MAGNET_OFFSET: 0 # TODO: Calibrate this on the final robot
-        BACK_RIGHT_MAGNET_OFFSET: 0 # TODO: Calibrate this on the final robot
-        GAZEBO_SIMULATION: False # Set to true if running in Gazebo
+    HALF_WHEEL_BASE: 0.7127 # HALF of the wheelbase in meters # TODO: Verify this on the final robot
+    HALF_TRACK_WIDTH: 0.36585 # HALF of the trackwidth in meters # TODO: Verify this on the final robot
```
Comment (pwmarshall) [inline config/drivetrain_config.yaml]: Verify/remove comment

File: src/motor_control/src/motor_control_node.cpp [hunk 2]
```diff
@@ -328,15 +336,15 @@ class MotorControlNode : public rclcpp::Node {
   void set_callback(const std::shared_ptr<rovr_interfaces::srv::MotorCommandSet::Request> request,
                     std::shared_ptr<rovr_interfaces::srv::MotorCommandSet::Response> response) {
 
-    if (request->type == "velocity") {
+    if (strcmp(request->type.c_str(), "velocity") == 0) {
```
Comment (jmblixt3) [inline src/motor_control/src/motor_control_node.cpp:342]: Strncmp is safer and reentrant

File: src/motor_control/src/motor_control_node.cpp [hunk 3]
```diff
@@ -218,11 +224,12 @@ class MotorControlNode : public rclcpp::Node {
     // Define default values for our ROS parameters below #
     this->declare_parameter("CAN_INTERFACE_TRANSMIT", "can0");
     this->declare_parameter("CAN_INTERFACE_RECEIVE", "can0");
-    this->declare_parameter("BACK_LEFT_TURN", 2);
-    this->declare_parameter("FRONT_LEFT_TURN", 4);
+    this->declare_parameter("BACK_LEFT_TURN", 4);
```
Comment (jmblixt3) [inline src/motor_control/src/motor_control_node.cpp:230]: Are default values needed, if not just leave them und set. If you leave this you'll need to change this value in multiple places

File: src/motor_control/src/motor_control_node.cpp [hunk 4]
```diff
@@ -240,17 +248,17 @@ class MotorControlNode : public rclcpp::Node {
         "motor/get", std::bind(&MotorControlNode::get_callback, this, _1, _2));
 
     // Instantiate all of our PIDControllers here
-    this->pid_controllers[this->get_parameter("BACK_LEFT_TURN").as_int()] = new PIDController(42, 0.01); // TODO: kp will need to be tuned on the real robot
-    this->pid_controllers[this->get_parameter("FRONT_LEFT_TURN").as_int()] = new PIDController(42, 0.01); // TODO: kp will need to be tuned on the real robot
-    this->pid_controllers[this->get_parameter("BACK_RIGHT_TURN").as_int()] = new PIDController(42, 0.01); // TODO: kp will need to be tuned on the real robot
-    this->pid_controllers[this->get_parameter("FRONT_RIGHT_TURN").as_int()] = new PIDController(42, 0.01); // TODO: kp will need to be tuned on the real robot
-    this->pid_controllers[this->get_parameter("SKIMMER_LIFT_MOTOR").as_int()] = new PIDController(42, 0.01); // TODO: kp will need to be tuned on the real robot
+    this->pid_controllers[this->get_parameter("BACK_LEFT_TURN").as_int()] = new PIDController(42, 0.002); // TODO: kp will need to be tuned on the real robot
```
Comment (jmblixt3) [inline src/motor_control/src/motor_control_node.cpp]: Can constants be extracted to config file

File: src/drivetrain/drivetrain/drivetrain_node.py [hunk 5]
```diff
@@ -260,13 +278,17 @@ def drive_callback(self, request, response):
...
         self.drive(msg.linear.y, msg.linear.x, msg.angular.z)
-        
+
     def absolute_encoders_callback(self, msg: AbsoluteEncoders) -> None:
         """This method is called whenever a message is received on the absoluteEncoders topic."""
-        self.front_left.current_absolute_angle = (360 * msg.front_left_encoder / 1023) - self.FRONT_LEFT_MAGNET_OFFSET
-        self.front_right.current_absolute_angle = (360 * msg.front_right_encoder / 1023) - self.FRONT_RIGHT_MAGNET_OFFSET
-        self.back_left.current_absolute_angle = (360 * msg.back_left_encoder / 1023) - self.BACK_LEFT_MAGNET_OFFSET
-        self.back_right.current_absolute_angle = (360 * msg.back_right_encoder / 1023) - self.BACK_RIGHT_MAGNET_OFFSET
+        front_left_adjusted = (msg.front_left_encoder - self.FRONT_LEFT_MAGNET_OFFSET) % 1023
```
Comment (jmblixt3) [inline src/drivetrain/drivetrain/drivetrain_node.py]: Can ticks be pulled out into config

**Review by jmblixt3 — COMMENTED** [review]
Here's some things to consider

File: src/motor_control/src/motor_control_node.cpp (hunk 3 above)
Comment (Isopod00) [reply src/motor_control/src/motor_control_node.cpp:230]: Yes, default values are needed. This is what is used when you "ros2 run" the node (no launch file) or when you launch the node without specifying a config file for the node.

File: src/drivetrain/drivetrain/drivetrain_node.py (hunk 5 above)
Comment (Isopod00) [reply src/drivetrain/drivetrain/drivetrain_node.py]: done

File: config/drivetrain_config.yaml (hunk 1 above)
Comment (Isopod00) [reply config/drivetrain_config.yaml]: We should measure this to get the exact values

File: src/motor_control/src/motor_control_node.cpp (hunk 3 above)
Comment (jmblixt3) [reply src/motor_control/src/motor_control_node.cpp:230]: Is that a good behavior though? I definitely see a see a scenario where you think you are updating a value but you aren't correctly updating it. If you see that as acceptable, then keep it, but looks like this mechanism could work, allow_undeclared_parameters

Comment (Isopod00) [reply src/motor_control/src/motor_control_node.cpp:230]: it's fine for now

File: config/drivetrain_config.yaml (hunk 1 above)
Comment (Isopod00) [reply config/drivetrain_config.yaml]: done

**Pull request merged**